rfc/redis-removal.md

```diff
@@ -0,0 +1,171 @@
+    primarily with the larger processes and timeline
+*   Identification of all new processes for maintaining services on a SQL
+    backend; these will be handled separately
+*   Prescription of what SQL vendor to use: Current functionality has been built
+    using MySQL; support for other vendors will be addressed separately
+
+## Case Against Redis
+
+Redis was introduced into the project as an easy way to get started. It's
+phenomenally flexible, and a single, common data store simplifies the
+operational story of the large operational footprint that Spinnaker has.
+Further, it was useful in that its flexibility meant services could be rapidly
+iterated on, as no schema was needed from the start.
+
+Unfortunately, Redis is difficult to operationalize as a persistent datastore,
+which many of the services treat Redis as. Community members are often surprised
+when Redis restarts to find Spinnaker has lost in-flight work. Operating Redis
+beyond a cache is not a skill that many organizations have. Worse yet, the level
+of effort to support a highly-available, persistent story for Redis is very high
+and littered with edge cases. For example, when tested at Netflix, both Redis
+Cluster and Sentinel regularly encountered incorrect master failover behavior
+that would lead to corrupted clusters or lost data.
+
+Furthermore, developing applications on Redis is typically a foreign concept and
+offers a confusing touch point for anyone looking to get involved in
+contributing.
+
+Having a singular datastore for Spinnaker is still necessary, as it eases the
+burden of onboarding. Now that Spinnaker is no longer a young project, the need
+for fast service iteration has largely dissapated. SQL was chosen as the
+replacement because modern SQL is flexible and well understood.
+
+## Service Inventory
+
+Most of the larger services already have support for SQL.
+
+* Full support: ✔️
+* Partial support: ~
+* No support: ✗
+
+| | |
+|-|-|
+| Clouddriver | ✔️ |
+| Echo | ✗ |
+| Fiat | ✗ |
+| Front50 | ✔️ |
+| Gate | ~ |
+| Igor | ✗ |
+| Kayenta | ~ |
+| Keel | ✔️ |
+| Orca | ✔️ |
+| Rosco | ✗ |
+| Swabbie | ✗ |
+
+## Service Migration Process
+
+_Some words on how services should be written such that they can be migrated
+without downtime._
+
+Being able to migrate existing deployments from Redis to SQL for each particular
+service will be crucial. Each service must provide these features and
+characteristics before they can be said to having "full support":
+
+* A service must support a dual-backend mode, where writes are sent to the SQL
+  store while reads can still hit Redis
+* A service must support an idempotent, background data migrator for all data
+  that needs to be migrated**
+* A service must provide a clear log signal of when migration is complete
+
+_** Caches and short-lived data do not need to be migrated_
+
+## Timeline
+
+This migration is assuming a slow migration cadence for the rest of the
+Spinnaker services.
+
+### 1.20 (early May)
```

Comment:
Does "complete technical tasks" imply that the service migration process listed above exists for the service? If so, does that process already exists for `clouddriver`, `front50`, `keel`, and `orca`?

Reply:
Yes, that is what it implies for all of those services - Keel excluded as it's only ever been MySQL. 

In fact, we've just tested the migration process again (SQL to SQL this time). These migration tools and processes are how we first moved away from Redis as well. Since we can't take downtime, we've been pretty deliberate about getting a good migration story in place for the services we've refactored so far.